### PR TITLE
Add `--listener` option to dogma watch command

### LINKED ISSUE #40: Feature Request: Add `--listener` option to `dogma watch` command
I'm wondering if `dogma` cli can have `--listener` option for `watch` subcommand.
What do you think?

# Example usage

```
$ cat listener.sh
#!/bin/sh

PATH=$1
TYPE=$2
REV=$3
URL=$4
MODIFIED_AT=$5
echo $PATH $REV $TYPE $URL $MODIFIED_AT
cat > $(basename $PATH)
kill -HUP $(pgrep prometheus)

$ dogma watch --listener listener.sh --streaming 'projFoo/repoA/*.yaml'
/foo.yaml 2 TEXT /api/v1/projects/projFoo/repos/repoA/contents/foo.yaml
/bar.yaml 4 TEXT /api/v1/projects/projFoo/repos/repoA/contents/bar.yaml
<Press CTRL+C>

$ ls
foo.yaml bar.yaml
```

# Background

I have configuration files (specifically, prometheus alert rule files) in a CentralDogma repository.
I'd like to fetch and store those config files and keep fetched-files up-to-date via watch API.
In addition, I need to send a SIGNAL to a certain process after fetching the latest content.
e.g. send kill HUP to prometheus process to let it reload config file

# Option details

Users can pass a command name to `--listener` option.
`dogma watch` command watches a CentralDogma repository and invokes given command with `watchResult`
parameters when it gets a notification. The user command will get updated content from STDIN and other
metadata from ARGV.

### PR DESCRIPTION
**Pull request by tom-lpsd**
Motivation:

Related to #40.

We're managing several middlewares (e.g. [prometheus](https://prometheus.io/), [HBase](https://hbase.apache.org/))
and we have had some configuration values in git repositories.

To make config-deploy process easier, we're going to put the config values to CentralDogma
so that each middleware host can fetch the latest config values quickly through watch API.

To keep config values on each middleware host in sync, we need a daemon application which
- runs on each middleware host
- watches target config values on CentralDogma repository
- synchronizes the local config file content with the updates on the repository

Of course, we can write a custom app using centraldogma library but
it'd be helpful if official cli tool can have a function to support such scenario.

Modifications:

- Add new option `--listener,-l` to dogma-watch command, that takes executable path as argument
- Add minimal test cases for the change

Result:

- Closes #40
- dogma cli users can handle centraldogma watch event by their custom script
  For example,

        dogma watch --listener myscript.sh --streaming /proj/repo/app_config.json

  `myscript.sh` gets watch event information through STDIN and environment variables like CGI.
  The content body of the watch target can be read from STDIN.
  The path, content_type, revision and url information can be accessed through env vars
  `DOGMA_WATCH_EVENT_PATH`, `DOGMA_WATCH_EVENT_CONTENT_TYPE`, `DOGMA_WATCH_EVENT_REV`, `DOGMA_WATCH_EVENT_URL` respectively.

Note:

In the issue #40, I propsed passing watch-event meta information through positional arguments (= $1, $2 ..) but I changed my mind.
If we adopt positional arguments, it might be difficult to keep compatibility when we change watch event structure.
That's why I use environment variables in this PR.

### CODE REVIEW

File: internal/app/dogma/watch_cmd.go [hunk 1]
```diff
@@ -30,14 +33,59 @@ type watchCommand struct {
 	repo      repositoryRequestInfo
 	jsonPaths []string
 	streaming bool
+	listener  string
```
Comment (minwoox) [inline internal/app/dogma/watch_cmd.go:36]: how about `listenerFile`?

**Review by minwoox — APPROVED** [review]
Looks good. Thanks!

**Review by minwoox — APPROVED** [review]
Looks good. Thanks!

File: internal/app/dogma/watch_cmd.go (hunk 1 above)
Comment (tom-lpsd) [reply internal/app/dogma/watch_cmd.go]: Thanks, that sounds better. updated 7203e94

Comment (tom-lpsd) [reply internal/app/dogma/watch_cmd.go]: oops, I forgot to add test change to the commit.
Here's amended fix cadda66.

File: internal/app/dogma/cli_cmds.go [hunk 2]
```diff
@@ -55,6 +55,11 @@ var streamingFlag = cli.BoolFlag{
 	Usage: "Specifies whether to keep watching the file",
 }
 
+var listenerFlag = cli.StringFlag{
+	Name:  "listener, l",
+	Usage: "Specifies the executable path that handles watch events",
```
Comment (ikhoon) [inline internal/app/dogma/cli_cmds.go:60]: Would you mind adding an explanation that shows what kinds of environment variables are injected for the listener?

**Review by ikhoon — APPROVED** [review]
Thanks @tom-lpsd 
Nice feature. I left a minor nit. :-)

File: internal/app/dogma/cli_cmds.go [hunk 3]
```diff
@@ -228,8 +228,21 @@ func CLICommands() []cli.Command {
...
+     DOGMA_WATCH_EVENT_CONTENT_TYPE - The content type of the file, JSON or TEXT
+     DOGMA_WATCH_EVENT_REV - The revision number of the watch event
+     DOGMA_WATCH_EVENT_URL - The URL of the target file
+
+   e.g.
+     # Print foo.json content when it gets updated
+     dogma watch --listener cat /pj/repo/foo.json`,
 			ArgsUsage: "<project_name>/<repository_name>/<path>",
 			Flags:     []cli.Flag{revisionFlag, jsonPathFlag, streamingFlag, listenerFlag},
 			Action: func(c *cli.Context) error {
```
Comment (ikhoon) [inline internal/app/dogma/cli_cmds.go]: nit: By the way, this file is mixed with tab and space though it is not related to this PR. Could you change the tabs to spaces, please?😉

File: internal/app/dogma/cli_cmds.go (hunk 2 above)
Comment (tom-lpsd) [reply internal/app/dogma/cli_cmds.go]: I hesitated to do that here because this "Usage" field seems to show short text about the flag.
Instead, I added "Description" section to the watch command c9b577e

Comment (ikhoon) [reply internal/app/dogma/cli_cmds.go]: It looks better. 👍

File: internal/app/dogma/cli_cmds.go (hunk 3 above)
Comment (tom-lpsd) [reply internal/app/dogma/cli_cmds.go]: Actually I tried to use tabs first but it broke help text indentation because `commandHelpTemplate` (https://github.com/line/centraldogma-go/pull/42/commits/c9b577e794349fa14c5acee334f67df96001b027#diff-33d65ee15203de0d14a825149497009c0697ad35c123552cd1a7f573c9958257L71) is indented using spaces.

like this
screen shot 2021-07-15 16 05 33

File: internal/app/dogma/watch_cmd.go [hunk 4]
```diff
@@ -55,37 +103,32 @@ func (wc *watchCommand) execute(c *cli.Context) error {
 	ctx, cancel := context.WithCancel(context.Background())
```
Comment (linxGnu) [inline internal/app/dogma/watch_cmd.go:103]: LGTM. 

Just a note that is not related to the PR:
- By looking at this file carefully, I think `ctx` here is no need.
- Done channel could be set to size 1 instead of 2. `1` sounds more `convention` to me.

**Review by linxGnu — APPROVED** [review]

File: internal/app/dogma/watch_cmd.go (hunk 4 above)
Comment (tom-lpsd) [reply internal/app/dogma/watch_cmd.go:103]: > I think ctx here is no need.

Exactly.

> 1 sounds more convention to me.

To me too.

I'll send another PR after this #42 gets merged.

**Comment by tom-lpsd** [comment]
Sorry but I realized that the test code in this PR doesn't work with old (< 1.14) go.

```
./watch_cmd_test.go:45:4: ts.EnableHTTP2 undefined (type *httptest.Server has no field or method EnableHTTP2)
```

So added 73c4025

PTAL, thanks.

By the way, travisCI is not working well in this repository it seems?

File: internal/app/dogma/cli_cmds.go (hunk 3 above)
Comment (ikhoon) [reply internal/app/dogma/cli_cmds.go]: I didn't know the details. Let's keep it as it is.

**Comment by minwoox** [comment]
> By the way, travisCI is not working well in this repository it seems?

Sorry about it. Let me add it. 😅

**Comment by tom-lpsd** [comment]
Thanks for the github action setup! I'll look into the test failure.

**Comment by minwoox** [comment]
Thanks @tom-lpsd!

**Comment by tom-lpsd** [comment]
> Could we hide the description of the listener in help when the OS is Windows?

Actually I tried to use the listener option on my windows machine and it worked. (I used windows `python` as executable for testing)
I guess `exec` package itself is portable enough to run on Windows.
The problem is that I don't come up with a portable way to write down a test that works on Windows.

I just added a short notice for windows users e7730f8.
What do you think?

**Comment by minwoox** [comment]
> Actually I tried to use the listener option on my windows machine and it worked. (I used windows python as executable for testing)
I guess exec package itself is portable enough to run on Windows.
The problem is that I don't come up with a portable way to write down a test that works on Windows.

Oops, I forgot this comment. Let me see what happened.

**Comment by minwoox** [comment]
@linxGnu Do you have a chance to look at why the test fails on Windows? 😅

**Comment by linxGnu** [comment]
Yeah let me take a look on Monday. I think of minor issue.

**Comment by tom-lpsd** [comment]
https://github.com/line/centraldogma-go/runs/3120517298
The error was a timeout and it didn't reproduce on my Windows machine.

In my environment, the skip call below just worked because there 's `cat` command on Windows.
https://github.com/line/centraldogma-go/pull/42/files#diff-9129127439518bbadf15d31891b135fd210c83fd1b38517a7e734641ffe98b8aR71

Somehow, Windows environment provided by Github Actions had `cat` command in search path but it got stuck when the test tried to run it.

**Comment by minwoox** [comment]
> The error was a timeout and it didn't reproduce on my Windows machine.

Yeah, I was wondering why it's timed out. 😅 
We can probably merge this PR and find out the cause later if you need this feature right away.
Do you need it?

**Comment by tom-lpsd** [comment]
I can wait for a while. Thank you.

**Comment by tom-lpsd** [comment]
The error was not specific to Windows but rather a bug I made by myself.
Here's the fix a7e33ea.

https://stackoverflow.com/questions/10473800/in-go-how-do-i-capture-stdout-of-a-function-into-a-string#comment46866218_29339052

As pointed out by this stackoverflow comment, read from a pipe should be done in another goroutine.
Otherwise, write call (`f()` call in my case) might block forever depending on the pipe buffer size.

**Comment by linxGnu** [comment]
@tom-lpsd IMHO, it's great fix and investigation ❤️ It's ok to merge for now.

However, after rethinking, I prefer non-hijack `os.Stdout`. It's race condition (at `os.Stdout`) if we run multiple tests in parallel.

How about `Dependency Injection` for central-dogma app which inject output upon instantiation. By DI, we are able to test and capture the output easily.

HDYT @minwoox @ikhoon @trustin

**Comment by minwoox** [comment]
> I prefer non-hijack os.Stdout. It's race condition (at os.Stdout) if we run multiple tests in parallel.

I agree with you on this.

> How about Dependency Injection for central-dogma app which inject output upon instantiation. By DI, we are able to test and capture the output easily.

So you mean, instead of using `os.Stdout`, we use a different output to test the result?
I'm not familiar with DI in Golang. 😅 Could you give me some advice or reference to it?

Anyway, I think we can merge this for now and make a separate issue to solve that. WDYT?

**Comment by linxGnu** [comment]
Yeah, I think we could merge and release ❤️ ❤️ 
It's great job @tom-lpsd 

Thank you so much

**Comment by minwoox** [comment]
Thanks, @tom-lpsd for your hard work. ❤️

**Comment by minwoox** [comment]
> I will try to apply DI in another PR

Thanks, @linxGnu. ❤️
